File: generated_code/GeminiPro/Python/008_SplayTree.py
import sys

# To prevent deep recursion errors on large trees
sys.setrecursionlimit(2000)

class SplayTree:
    """
    An implementation of a Splay Tree that acts as a dictionary-like set for integers.

    A splay tree is a self-adjusting binary search tree with the additional
    property that recently accessed elements are quick to access again.
    It performs basic operations such as insertion, look-up and removal
    in O(log n) amortized time.
    """

    class _Node:
        """A node in the Splay Tree."""
        def __init__(self, key):
            self.key = key
            self.parent = None
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x: _Node):
        """Performs a left rotation on the given node x."""
        y = x.right
        x.right = y.left
        if y.left:
            y.left.parent = x
        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        y.left = x
        x.parent = y

    def _right_rotate(self, x: _Node):
        """Performs a right rotation on the given node x."""
        y = x.left
        x.left = y.right
        if y.right:
            y.right.parent = x
        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
        y.right = x
        x.parent = y

    def _splay(self, x: _Node):
        """
        Performs the splaying operation on node x, moving it to the root.
        """
        while x.parent:
            parent = x.parent
            grandparent = parent.parent
            if not grandparent:  # Zig case
                if x == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif x == parent.left and parent == grandparent.left:  # Zig-Zig case (left-left)
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif x == parent.right and parent == grandparent.right:  # Zig-Zig case (right-right)
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            elif x == parent.right and parent == grandparent.left:  # Zig-Zag case (left-right)
                self._left_rotate(parent)
                self._right_rotate(grandparent)
            else:  # Zig-Zag case (right-left)
                self._right_rotate(parent)
                self._left_rotate(grandparent)

    def search(self, key: int) -> bool:
        """
        Searches for a key in the tree.

        If the key is found, the corresponding node is splayed to the root.
        If the key is not found, the last accessed node (the parent of the
        would-be key) is splayed to the root.

        Args:
            key: The integer key to search for.

        Returns:
            True if the key is found, False otherwise.
        """
        if not self.root:
            return False

        current = self.root
        last_visited = None
        while current:
            last_visited = current
            if key == current.key:
                self._splay(current)
                return True
            elif key < current.key:
                current = current.left
            else:
                current = current.right
        
        # If key not found, splay the last visited node
        if last_visited:
            self._splay(last_visited)
            
        return False

    def insert(self, key: int):
        """
        Inserts a key into the Splay Tree.

        If the key already exists, the existing node is splayed to the root.
        If the key is new, it is inserted and the new node is splayed to the root.

        Args:
            key: The integer key to insert.
        """
        if not self.root:
            self.root = self._Node(key)
            return

        current = self.root
        parent = None
        while current:
            parent = current
            if key == current.key:
                # Key already exists, splay it and return
                self._splay(current)
                return
            elif key < current.key:
                current = current.left
            else:
                current = current.right

        new_node = self._Node(key)
        new_node.parent = parent
        if key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node
        
        self._splay(new_node)

    def delete(self, key: int):
        """
        Deletes a key from the Splay Tree.

        The tree is re-structured after deletion to maintain the splay tree properties.
        Specifically, the parent of the deleted node is splayed to the root.

        Args:
            key: The integer key to delete.
        """
        # Search for the key, which will splay it (or a nearby node) to the root
        if not self.search(key):
            # Key not in tree, nothing to do
            return

        # After search(key), if the key was found, it's now the root
        # If self.root.key is not key, it means the key was not found.
        if self.root.key != key:
            return

        node_to_delete = self.root
        left_subtree = node_to_delete.left
        right_subtree = node_to_delete.right

        if not left_subtree:
            self.root = right_subtree
            if right_subtree:
                right_subtree.parent = None
        elif not right_subtree:
            self.root = left_subtree
            if left_subtree:
                left_subtree.parent = None
        else:
            # Disconnect left subtree from the node to be deleted
            left_subtree.parent = None
            
            # Find the maximum element in the left subtree
            max_node = left_subtree
            while max_node.right:
                max_node = max_node.right
            
            # Splay this maximum element to the top of the left subtree
            # This is a splay operation on a detached tree, so we temporarily
            # set self.root to the root of the left subtree.
            temp_root = self.root
            self.root = left_subtree
            self._splay(max_node) # Now self.root is max_node
            
            # Attach the original right subtree to the new root
            self.root.right = right_subtree
            right_subtree.parent = self.root
            
            # Restore the original root if it was changed for the splay
            # (although in this logic, self.root is correctly set now)

        # Clean up the deleted node's references to be safe
        del node_to_delete

    def __str__(self):
        """Returns an in-order traversal string representation of the tree."""
        if not self.root:
            return "[]"
        
        result = []
        def inorder_traverse(node):
            if node:
                inorder_traverse(node.left)
                result.append(str(node.key))
                inorder_traverse(node.right)
        
        inorder_traverse(self.root)
        return "[" + ", ".join(result) + "]"

File: generated_code/GeminiPro/Java/008_SplayTree.java
/**
 * A complete, self-contained Java class that implements an integer set using a Splay Tree.
 * A splay tree is a self-adjusting binary search tree with the additional
 * property that recently accessed elements are quick to access again.
 * It performs basic operations such as insertion, deletion, and search.
 * After a search, insert, or delete operation, the accessed node is moved
 * to the root of the tree (an operation called "splaying").
 */
public class SplayTree {

    /**
     * Represents a node in the Splay Tree.
     * It is a private static inner class as it is tightly coupled with the SplayTree
     * and does not need access to SplayTree's instance members.
     */
    private static class Node {
        int key;
        Node left;
        Node right;

        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * Performs a right rotation on the given node y.
     *
     *      y           x
     *     / \         / \
     *    x   T3  ->  T1  y
     *   / \             / \
     *  T1  T2          T2  T3
     *
     * @param y The node to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node rightRotate(Node y) {
        Node x = y.left;
        y.left = x.right;
        x.right = y;
        return x;
    }

    /**
     * Performs a left rotation on the given node x.
     *
     *    x             y
     *   / \           / \
     *  T1  y   ->    x   T3
     *     / \       / \
     *    T2  T3    T1  T2
     *
     * @param x The node to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The main splaying operation. It brings the node with the given key
     * (or the last accessed node if the key is not found) to the root of the tree.
     * This is a recursive, bottom-up splay implementation.
     *
     * @param node The root of the subtree to splay.
     * @param key The key to search for and bring to the root.
     * @return The new root of the splayed subtree.
     */
    private Node splay(Node node, int key) {
        // Base case: node is null or key is at root
        if (node == null || node.key == key) {
            return node;
        }

        // Key lies in the left subtree
        if (key < node.key) {
            // Key is not in the tree, we are done
            if (node.left == null) {
                return node;
            }

            // Zig-Zig (Left-Left)
            if (key < node.left.key) {
                node.left.left = splay(node.left.left, key);
                node = rightRotate(node);
            }
            // Zig-Zag (Left-Right)
            else if (key > node.left.key) {
                node.left.right = splay(node.left.right, key);
                if (node.left.right != null) {
                    node.left = leftRotate(node.left);
                }
            }
            // Do second rotation for root
            return (node.left == null) ? node : rightRotate(node);
        }
        // Key lies in the right subtree
        else {
            // Key is not in the tree, we are done
            if (node.right == null) {
                return node;
            }

            // Zig-Zag (Right-Left)
            if (key < node.right.key) {
                node.right.left = splay(node.right.left, key);
                if (node.right.left != null) {
                    node.right = rightRotate(node.right);
                }
            }
            // Zig-Zig (Right-Right)
            else if (key > node.right.key) {
                node.right.right = splay(node.right.right, key);
                node = leftRotate(node);
            }
            // Do second rotation for root
            return (node.right == null) ? node : leftRotate(node);
        }
    }

    /**
     * Searches for a key in the tree. If the key is found, it is splayed
     * to the root. If not found, the last accessed node is splayed to the root.
     *
     * @param key The key to search for.
     * @return The key as an Integer object if found, otherwise null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }
        root = splay(root, key);
        // After splaying, if the key exists, it will be at the root.
        if (root.key == key) {
            return key;
        }
        return null;
    }

    /**
     * Inserts a key into the splay tree. If the key already exists,
     * the tree is splayed on that key but not modified, maintaining the set property.
     *
     * @param key The key to insert.
     */
    public void insert(int key) {
        // If the tree is empty, create a new root
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree to bring the closest node to the root
        root = splay(root, key);

        // If key is already present, do nothing
        if (root.key == key) {
            return;
        }

        Node newNode = new Node(key);

        // The new node becomes the new root. The old root and its subtrees
        // are attached as children of the new root.
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Deletes a key from the splay tree. If the key is not found,
     * the tree is splayed on the last accessed node, but no deletion occurs.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the tree to bring the node to delete (if it exists) to the root
        root = splay(root, key);

        // If the key is not in the tree, do nothing
        if (root.key != key) {
            return;
        }

        // Now, the node to delete is the root. We need to merge its two subtrees.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If there's no left subtree, the right subtree becomes the new tree
            root = rightSubtree;
        } else {
            // Make the left subtree the new tree
            root = leftSubtree;
            
            // Find the maximum element in the new tree (the rightmost node of the left subtree)
            // and splay it to the root of this new tree.
            Node maxNode = root;
            while (maxNode.right != null) {
                maxNode = maxNode.right;
            }
            root = splay(root, maxNode.key);

            // After splaying, the new root (the max element from the left subtree)
            // has no right child. We can now attach the original right subtree there.
            root.right = rightSubtree;
        }
    }
}